### DISCUSSION TITLE
QRCodeException - code length overflow

### DISCUSSION THREAD

**Discussion by mira-kovar** [chillerlan/php-qrcode#187] [General]
Hi,
I have problem with generating SVG QRCode. No matter which size I'm gettting `QRCodeException` Here is my content. It is valid QRCode content for Czech standart for QR Invoice (http://qr-faktura.cz/).

Please could you increase `MAX_BITS` or provide option to bypase?

```
SPD*1.0*ACC:CZ8620100000002000948394*AM:3025.00*CC:CZK*X-VS:1181273*X-SS:0*X-KS:308*MSG:1181273*DT:20190318*X-INV:SID%2A1.0%2ATB0%3A2500.00%2AT0%3A525.00%2AID%3A1181273%2ADD%3A20190304%2AINI%3A25520270%2AVII%3ACZ25520270%2AINR%3A24791334%2AVIR%3ACZ24791334%2ADUZP%3A20190304%2AAM%3A3025.00%2AVS%3A1181273%2ADT%3A20190318%2ACC%3ACZK%2AX-SW%3Acpilot.cz*
```

**Comment by codemasher** [comment]
I just tried the data you posted with the SVG example and it worked perfectly fine with `QRCode::VERSION_AUTO`. What is the value for [`QROptions::$version`](https://github.com/chillerlan/php-qrcode/blob/master/src/QROptionsTrait.php#L17-L24) you're using? The minimum (for 2822 bits) would be 12 for ECC type L according to [`QRDataInterface::MAX_BITS`](https://github.com/chillerlan/php-qrcode/blob/ec8e534ee941369e2e1e439ef85fbf38364699d1/src/Data/QRDataInterface.php#L78-L120). See http://www.qrcode.com/en/about/version.html for more info on that.

**Comment by mira-kovar** [comment]
Aha! I used version 5, I works like charm. Thanks

**Comment by MichaMeyer** [comment]
Hi! We have a similar problem!
1.) We pass a URL for qr code creation: 
https://specs.abc-efg.com/Administrator/

it has a length of 40 Characters, we have chosen to use version 4 QR Code and ECC H
Specification says with ECC H **50 alphanumeric letters** are possible? 
![grafik](https://user-images.githubusercontent.com/32386846/226602628-cc47ae50-46fb-4210-81f0-2ddf87dfb3ca.png)

we get following error: `Uncaught chillerlan\QRCode\Data\QRCodeDataException: code length overflow. (292 > 288 bit)`

2.) My developers told me that the newest version of the library would not work, and that it was necessary to "combine" several versions of the lib and make own extensions in the lib?

They worked based on this:
https://github.com/chillerlan/php-qrcode/blob/main/examples/imageWithLogo.php

We just needed a normal QR Code with an Logo (with logospace around) inside and 

- variable colors for foreground, 
- background and 
- Background transparency. 

Is there any problem to create this with the latest version? 

And which one (stable or not?) should be used?

3.) We would like to use Version 4 + ECC Q with the logo to have more possible letters in the url. This works at a lot QR Generators in the web, but we get an errormessage "that Logo is only allowed with ECC H used"  Is this possible in any way?

Would really appreciate your help!

**Reply by codemasher** [reply]
An URL is not "Alphanumeric" as this mode only works with UPPERCASE letters. the above example URL will not work in that mode at all. The characters in this map are allowed: 
https://github.com/chillerlan/php-qrcode/blob/d4887f267641d8fed151069fdf35c29fab252a48/src/Data/AlphaNum.php#L30-L37

The internal restriction to ECC level H is due to the fact that the logo space uses up most of the error correction bits, to guarantee that the QR Code remains readable. If I'd remove the restriction I'd probably get countless complaints...
That said, I could make the ECC requirement level optional? ECC H by default or whatever you set in your own application at your own risk?

**Reply by MichaMeyer** [reply]
thanks for the fast answer! 

so how can we find out how long the name after the slash e.g. "Administrator" is allowed to be, to be coded in Version 4 QR? 
Anything out of the specs I attached?

And about question 2 I asked: 
Are the things I described possible with the actual library, and my programmers just have not solved it correctly (on base of your example file)  or are there bugs inside who prevent the functionality we need? 

Its just to be sure because they are external developers and I can not find out if they just not able to do it, or if it is currently not possible?

**Reply by codemasher** [reply]
Honestly i don't really know what they mean by "combine" things. Is it possible to direct them over here so they can tell me directly what they're trying to do?
As for the character length: it is safest to assume the character lengths for "binary" mode as URL paths and parameters can be case sensitive.
Also, may i ask why a version 4 QR Code specifically?

**Reply by MichaMeyer** [reply]
okay, they told me that they have to "combine different releases of your lib" to make it work.. 

We need to put the qr code as PNG with always the same size e.g. 1800x1800px on a product. so we thought this is only possible when using always the same Version (in this case 4)? Is this correct? 

Also it should be readable from different angles and bigger distances. so we assumed less modules (version4 instead of 5) will increase the probability of detection?

**Reply by codemasher** [reply]
Hey, I'm still wondering why they have to "combine different releases". I understand that the current `dev-main` may not be an option for production yet, however, I'm close to finishing v5 and it should be ready to use soon.

About the size of the QR Code there seem to be some misconceptions:

- The final size (e.g. the image is upscaled on) has nothing to do with the QR Code version. In fact, if you choose a lower version, there are fewer remaining error correction bits (assuming the same logo space size at a higher version) - this is especially important when the QR Code is printed and there is a possibility it could be damaged.
- Modern QR Code readers are capable of reading from a 4K screen [even 1 pixel sized QR modules](https://github.com/chillerlan/php-qrcode/blob/main/tests/samples/smol.png) (`QROptions::$scale = 1`), so a printed QR Code at e.g. 10x10 mm and version 5 would work perfectly fine.

**Reply by MichaMeyer** [reply]
Hi Codemasher! As you are from Germany as I , is there a possibility to communicate differently as here in community? 

You already have answered several questions to my Devs regarding "transparency problems when scaling..."

From my side I can not imagine that this is a problem of your library? Could qr code loose transparency when getting scaled in PHP to a PNG in fixed size?

**Reply by codemasher** [reply]
Hey, i think it is the same issue as this one: https://github.com/chillerlan/php-qrcode/discussions/190#discussioncomment-5453785 (transparency needs to be re-applied after the scaling operation). This is not necessarily an issue of this library but more how the GD extension works internally.

You may contact me via my email address that is plastered all over GitHub, however, the preferred way solving code issues is this forum.

**Reply by codemasher** [reply]
It's in the header of every single source file I created:
https://github.com/chillerlan/php-qrcode/blob/284c42db884b98b2453e39a7e10a7a04239d0019/src/QRCode.php#L6

**Reply by MichaMeyer** [reply]
okay, sorry!

**Reply by MichaMeyer** [reply]
sended.. Thank you!
